### PR TITLE
Add in-memory chunk cache

### LINKED ISSUE #6239: Store Gateway: Inmemory chunks cache
**Is your feature request related to a problem? Please describe.**
For scenarios where rules are configured to query Store Gateway, Store Gateway can query the same chunks over and over again.

Existing chunk cache only supports memcached and Redis. Fetching the same data over network could increase the bandwidth usage baseline, which is not ideal

**Describe the solution you'd like**
Add a new inmemory chunks cache backend so that chunks can be stored inmemory.

### PR DESCRIPTION
**Pull request by SungJin1212**
Add an in-memory chunk cache to Store Gateway.

**Which issue(s) this PR fixes**:
Fixes #6239 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,10 +2,11 @@
 
 ## master / unreleased
 
-* [FEATURE] Query Frontend/Querier: Add protobuf codec `-api.querier-default-codec` and the option to choose response compression type `-querier.response-compression`. #5527
 * [CHANGE] Enable Compactor and Alertmanager in target all. #6204
+* [FEATURE] Query Frontend/Querier: Add protobuf codec `-api.querier-default-codec` and the option to choose response compression type `-querier.response-compression`. #5527
```
Comment (SungJin1212) [inline CHANGELOG.md:6]: I just changed order

**Comment by SungJin1212** [comment]
@yeya24 
How about setting an in-memory cache as a default?

**Comment by yeya24** [comment]
@SungJin1212 I prefer to stick to current behavior to not enable chunks caching as default

**Comment by SungJin1212** [comment]
@yeya24 
I have updated the pr.

**Review by yeya24 — APPROVED** [review]
Awesome work! Thanks for the contribution

**Review by friedrichg — APPROVED** [review]

**Pull request merged**